compiler/src/dotty/tools/dotc/eff/EffectOps.scala

```diff
@@ -0,0 +1,203 @@
+        case _ => tp
+
+    def getKilled(using Context): List[Tree] =
+      tp match
+        case KillType(_, killSet) => killSet
+        case _ => Nil
+
+    def isKillFun(using Context): Boolean =
+      tp match
+        case fntpe @ FunctionOrMethod(_, KillType(_, _)) => true
+        case _ => false
+
+    def isKillType(using Context): Boolean =
+      tp match
+        case KillType(_) => true
+        case _ => false
+
+  /**
+   * Checks that the variables inside a kill annotation are well-formed
+   * Well-formedness conditions:
+   * 1. Must be non-empty
+   * 2. No duplicates allowed e.g. no kill(f, f).
+   * 3. Must be a capture trackable ref
+   * 4. Must not be a special capability
+   * 5. Kill annotation can only appear in (NOT YET CHECKED! TODO!)
+   *  a) At top-level of explicit DefDef tpt
+   *  b) As result type of a MethodType
+   */
+  def checkWellformed(annot: Tree)(using Context): Unit =
+    val killedElems = annot.killedElems
+    if killedElems.isEmpty then
+      report.error(i"Kill set of $annot may be empty.", annot.srcPos)
+    else
+      val killSet = util.HashSet[Symbol]()
+      for ref <- killedElems do
+        val refSym = ref.symbol
+        if !killSet.add(refSym) then
+          report.error(i"Kill set of $annot has a duplicate ref $ref", annot.srcPos)
+        ref.tpe match
+          case ref: CaptureRef if ref.isTrackableRef =>
+            if ref.isRootCapability then // hopefully only case that needs handling.
+              report.error(i"Killed variable cannot be a root capability!", annot.srcPos)
+          case _ =>
+            report.error(i"Killed variable ${ref} is not a capability!", annot.srcPos)
+  end checkWellformed
+end KillOps
+
+object EffectType:
+  import EffOps.*
+
+  def apply(tp: Type, usedRefs: List[Tree], killedRefs: List[Tree])(using Context): Type =
+    val annotTree =
+      Apply(
+        New(getEffAnnot.typeRef,
+          Typed(
+            SeqLiteral(usedRefs, TypeTree(defn.AnyType)),
+            TypeTree(defn.RepeatedParamClass.typeRef.appliedTo(defn.AnyType))
+          ) :: Nil),
+        Typed(
+          SeqLiteral(killedRefs, TypeTree(defn.AnyType)),
+          TypeTree(defn.RepeatedParamClass.typeRef.appliedTo(defn.AnyType))
+        ) :: Nil
+      )
+    AnnotatedType(tp, Annotation(annotTree))
+
+  def unapply(tp: Type)(using Context): Option[(Type, List[Tree], List[Tree])] =
+    tp match
+      case AnnotatedType(parent, annot) if annot.symbol.isEff =>
+        Some(parent, annot.tree.usedElems, annot.tree.killedElems)
+      case _ => None
+end EffectType
+
+object EffOps:
+  extension (tree: Tree)
+    def usedElems: List[Tree] = tree match
+      case Apply(
+        Apply(_, Typed(SeqLiteral(elems, _), _) :: Nil), _
+        ) => elems
+      case _ => Nil
+
+    // This is the same as KillOps.killedElems
+    def killedElems: List[Tree] = tree match
+      case Apply(_, Typed(SeqLiteral(elems, _), _) :: Nil) => elems
+      case _ => Nil
```

Comment:
Such `...Elems` methods seems generic, depending only on the annotation structure, and they themselves do not check for whether it is kill or effect. Perhaps could merge them and leave a note referring to `...Type.apply`.